### ISSUE TITLE
How do we want to handle typos during scraping?

### ISSUE THREAD

**Issue by terinpw** [Recidiviz/recidiviz-data#789]
Was originally thinking this in terms of enum strings (any typo right now will a scrape task to fail), but typo handling could be applied more broadly (for instance entity matching)

**Comment by jessex** [comment]
Do you mean typos in the source data itself or typos in our data extraction or something else?

**Comment by jamwalla** [comment]
I think this is about typos in the source data. This is related to #355, I think with the idea that some typos will cause non-"catastrophic" errors.

We've been concerned about biasing the data if we fail quietly in some cases. For example, if a jail rarely has people designated Asian in custody, and then we discover that they're using a string we don't handle in conversion, we don't want to quietly drop these people - we perhaps want to fail loudly, invalidating the whole scrape session. But this is probably not what we want to do if we can't parse a charge class, for example.

I think Terin brought this up because brooks jeffery rosters have manual entry (!?) for race and gender, and they get fat-fingered on occasion. If we don't want to fail in these cases, and if we still want to entity match, we need to be resilient to errors in the source.

**Comment by terinpw** [comment]
Yup, +1 to what @jamwalla wrote, sorry for the lack of context. 

For concrete examples - one person page in a brooks jeffrey region had the text 'Bo bond Available' instead of 'No bond available', which throws an error because it doesn't map to any BondStatus. If we end up supporting some type of fuzzy matching for enum mappings, this could be handled. 

Similarly, there was another person in a brooks jeffrey region whose race and gender fields were reversed (Gender was W and Race was M -- instead of Gender being M(ale) and race being W(hite), which conforms with the rest of the website). I don't think we can really predict something like this, but we probably shouldn't invalidate the entire scrape session because of something like this...

Originally with mentioning entity matching - I was thinking that we might run into similar issues with a person's name being mistyped and therefore not matched correctly to an existing entity. While this could be a risk, it's probably not a big one since we aren't doing any person matching across different bookings unless there are explicit ids (I'm making the assumption here that within a single booking, a person isn't going to be re-entered into the system manually). 


I think there are 2 parts to this issue: 

1) Are there any situations where we want to fuzzy match to handle typos?

2) What invalidates a scrape session? Is it 1 error? A % of failed tasks? Is there a subset of errors that we never want to invalidate a scrape session for?

**Comment by jessex** [comment]
This is fantastic context! I would very much like for fuzzy matching to be incorporated into ingest. I confess that I have never worked directly with fuzzy matching tech myself, but happy to come up to speed to help review ideas and code. Ideally there would be some sort of drop-in library that can be "trained" (not in an ML sense necessarily) on / configured for the kinds of scenarios we want to look for. For example, I think the edge case of similar looking but distinct fields sometimes having their values reversed is something that is non-standard.

More specific thoughts on @terinpw's two questions:
1. Assuming we can quantify and bound are error rates, I think we'd want to fuzzy match on identifiable fields, e.g. demographics, and potentially also on non-identifiable, low-variability fields like some of our enum fields. The thought on the first is that it's probably more valuable to fuzzy match with confidence on those fields to support calculation than to just give up on classes of metrics in affected jurisdictions. The thought on the second is that these are the kinds of fields that more complex logic is likely to be built atop, and fuzzy matching into discrete values should be easier in theory. But again, not an expert on fuzzy matching at all yet.
2. I think I remember Arian or Colin speaking about this earlier on as regards monitoring that we want to come up with a classification of error types, e.g. that some are permanent/catastrophic and some are transient/retryable, and some are session-side and some are individual-specific. If we had such a classification, we could come up with error types in the code that would trigger different failure modes.
